### ISSUE TITLE
cant save portal2 maps: Failed to convert PLANES to bytes!

### ISSUE THREAD

**Issue by sum-catnip** [snake-biscuits/bsp_tool#112]
Im trying to load and save portal2 maps:
```python
import bsp_tool as bsp
m = bsp.load_bsp('sp_a1_intro1.bsp', bsp.branches.valve.sdk_2013)
print(m)
m.save()
```

which leads to this error:
```python
λ python app.py
<ValveBsp 'sp_a1_intro1.bsp' valve.sdk_2013 (VBSP version 21)>
Failed to convert PLANES to bytes!
Traceback (most recent call last):
  File "/home/catnip/repos/portal2/mapgen/app.py", line 9, in <module>
    m.save()
  File "/home/catnip/venv/mapgen/lib/python3.11/site-packages/bsp_tool/base.py", line 111, in save
    self.save_as(os.path.join(self.folder, self.filename))
  File "/home/catnip/venv/mapgen/lib/python3.11/site-packages/bsp_tool/valve.py", line 137, in save_as
    raise exc
  File "/home/catnip/venv/mapgen/lib/python3.11/site-packages/bsp_tool/valve.py", line 134, in save_as
    lump_bytes = self.lump_as_bytes(LUMP.name)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/catnip/venv/mapgen/lib/python3.11/site-packages/bsp_tool/valve.py", line 116, in lump_as_bytes
    raw_lump = b"".join([struct.pack(_format, *x.as_tuple()) for x in lump_entries])
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/catnip/venv/mapgen/lib/python3.11/site-packages/bsp_tool/valve.py", line 116, in <listcomp>
    raw_lump = b"".join([struct.pack(_format, *x.as_tuple()) for x in lump_entries])
                                               ^^^^^^^^^^^^
  File "/home/catnip/venv/mapgen/lib/python3.11/site-packages/bsp_tool/branches/base.py", line 218, in as_tuple
    return [int(x) if f in "bBhHiI" else x for x, f in zip(_tuple, split_format(self._format))]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/catnip/venv/mapgen/lib/python3.11/site-packages/bsp_tool/branches/base.py", line 218, in <listcomp>
    return [int(x) if f in "bBhHiI" else x for x, f in zip(_tuple, split_format(self._format))]
            ^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'PlaneType'
```

i tried with and without specifying the branch.

**Comment by snake-biscuits** [comment]
This could take a while to "fix" since it requires every `.as_bytes()` method functions correctly
So far I'm fairly confident I've fixed the `.as_bytes()` case you got, which affected a subset of `LumpClasses`

All that should be left is `SpecialLumpClasses`, which are mostly untested.
`Visibility` is currently busted, seems to be a problem with compression
`PhysicsCollide` is also very broken, and has been for a while

Popping lumps from `valve.sdk_2013.SPECIAL_LUMP_CLASSES` will ensure they save correctly
(Since the raw bytes of the lump are copied, rather than recreated via the `.from_bytes -> .as_bytes` pipeline)
You can do this without editing the branch script, if you like:
```python
from bsp_tool import ValveBsp
from bsp_tool.branches.valve import sdk_2013

sdk_2013.SPECIAL_LUMP_CLASSES.pop("PHYSICS_COLLIDE")
bsp = ValveBsp(sdk_2013, "path/filename.bsp")
...
bsp.save_as(...)
```
But unparsed lumps cannot be edited, so if you wanted to do that you'll have to wait for fixes.
Also, `bsp_tool` will still save all lumps as decompressed copies of the original
(I should probably look at allowing users to re-compress, maybe)

**Comment by sum-catnip** [comment]
very much appreciated :)

**Comment by sum-catnip** [comment]
its already looking alot better. i tried calling asbytes on all lumps with a try-catch and this is what i got:

```
ENTITIES
PLANES
TEXTURE_DATA
VERTICES
VISIBILITY
NODES
TEXTURE_INFO
pack expected 18 items for packing (got 17)
FACES
LIGHTING
OCCLUSION
LEAVES
pack expected 15 items for packing (got 14)
FACE_IDS
EDGES
SURFEDGES
MODELS
WORLD_LIGHTS
LEAF_FACES
LEAF_BRUSHES
BRUSHES
BRUSH_SIDES
AREAS
AREA_PORTALS
FACE_BRUSHES
FACE_BRUSH_LIST
UNUSED_24
UNUSED_25
DISPLACEMENT_INFO
ushort format requires 0 <= number <= 65535
ORIGINAL_FACES
PHYSICS_DISPLACEMENT
PHYSICS_COLLIDE
invalid model type
VERTEX_NORMALS
VERTEX_NORMAL_INDICES
DISPLACEMENT_LIGHTMAP_ALPHAS
DISPLACEMENT_VERTICES
DISPLACEMENT_LIGHTMAP_SAMPLE_POSITIONS
GAME_LUMP
argument for 's' must be a bytes object
LEAF_WATER_DATA
PRIMITIVES
PRIMITIVE_VERTICES
PRIMITIVE_INDICES
PAKFILE
CLIP_PORTAL_VERTICES
CUBEMAPS
TEXTURE_DATA_STRING_DATA
TEXTURE_DATA_STRING_TABLE
OVERLAYS
LEAF_MIN_DIST_TO_WATER
FACE_MACRO_TEXTURE_INFO
DISPLACEMENT_TRIS
PROP_BLOB
WATER_OVERLAYS
LEAF_AMBIENT_INDEX_HDR
LEAF_AMBIENT_INDEX
LIGHTING_HDR
WORLD_LIGHTS_HDR
LEAF_AMBIENT_LIGHTING_HDR
LEAF_AMBIENT_LIGHTING
XZIP_PAKFILE
FACES_HDR
MAP_FLAGS
OVERLAY_FADES
OVERLAY_SYSTEM_LEVELS
PHYSICS_LEVEL
DISPLACEMENT_MULTIBLEND
```

the first one is TEXTURE_INFO. i think the problem is that this source engine version uses vec4s intead of vec3s for lightmap and texture. taken from the 2018 source engine leak:

```c
typedef struct texinfo_s
{
	DECLARE_BYTESWAP_DATADESC();
	float		textureVecsTexelsPerWorldUnits[2][4];			// [s/t][xyz offset]
	float		lightmapVecsLuxelsPerWorldUnits[2][4];			// [s/t][xyz offset] - length is in units of texels/area
	int			flags;				// miptex flags + overrides
	int			texdata;			// Pointer to texture name, size, etc.
} texinfo_t;
``` 

and this is how dleaf_t looks:

```c
struct dleaf_version_0_t
{
	DECLARE_BYTESWAP_DATADESC();
	int				contents;			// OR of all brushes (not needed?)

	short			cluster;

	BEGIN_BITFIELD( bf );
	short			area:9;
	short			flags:7;			// Per leaf flags.
	END_BITFIELD();

	short			mins[3];			// for frustum culling
	short			maxs[3];

	unsigned short	firstleafface;
	unsigned short	numleaffaces;

	unsigned short	firstleafbrush;
	unsigned short	numleafbrushes;
	short			leafWaterDataID; // -1 for not in water

	// Precaculated light info for entities.
	CompressedLightCube m_AmbientLighting;
};

// version 1
struct dleaf_t
{
	DECLARE_BYTESWAP_DATADESC();
	int				contents;			// OR of all brushes (not needed?)

	short			cluster;

	BEGIN_BITFIELD( bf );
	short			area:9;
	short			flags:7;			// Per leaf flags.
	END_BITFIELD();

	short			mins[3];			// for frustum culling
	short			maxs[3];

	unsigned short	firstleafface;
	unsigned short	numleaffaces;

	unsigned short	firstleafbrush;
	unsigned short	numleafbrushes;
	short			leafWaterDataID; // -1 for not in water

	// NOTE: removed this for version 1 and moved into separate lump "LUMP_LEAF_AMBIENT_LIGHTING" or "LUMP_LEAF_AMBIENT_LIGHTING_HDR"
	// Precaculated light info for entities.
//	CompressedLightCube m_AmbientLighting;
};
```

is uses one of those based on the leaf lump version (version 0 or 1).

maybe this helps :) if you want i can also give you my hexpat file which has about 20 working structs working with portal2 maps

**Comment by snake-biscuits** [comment]
~~Looks like I fixed all the lumps you found errors with while working on this issue~~
**EDIT**: `GameLump_sprp.as_bytes()` is broken, not tested either
At the very least lots has been fixed since the 0.4.0 PyPI release

You might have to make sure you're using the latest version of `bsp_tool` from GitHub
Shipping a new version to PyPI might be a ways off yet, since there's breaking changes
`extensions.diff` is also halfway through a refactor so I can automate testing for saves
Testing for the major refactors to how changes are made to loaded bsps is also incomplete

Focusing in on saving is definitely a priority rn
I'll aim to keep v`0.5.0` to saving changes, without too many extras.

**Comment by sum-catnip** [comment]
Oh i pip installed from the master branch yesterday so im pretty sure i was using the latest version

**Comment by snake-biscuits** [comment]
That's surprising, I'm not sure why `as_bytes` is breaking then, it looks like some values are skipped / misaligned from the error messages

Are you trying to port maps across branches? that could explain some things if so

**Comment by snake-biscuits** [comment]
AFAIK `float[2][4]` TextureVectors aren't new, they're how Hammer stores texture coordinates in `.vmf` files

**Comment by snake-biscuits** [comment]
TextureInfo doesn't seem to have changed since `valve.source`[^ti]
Same with Leaf version 0[^L0] & 1[^L1] (Leaf v1 is defined in `valve.orange_box`)

[^ti]: 2013 Source SDK `texinfo_s` [`src/public.bspfile.h` line 497](https://github.com/ValveSoftware/source-sdk-2013/blob/master/sp/src/public/bspfile.h#L497)
[^L0]: 2013 Source SDK `dleaf_version_0_t` [`src/public.bspfile.h` line 796](https://github.com/ValveSoftware/source-sdk-2013/blob/master/sp/src/public/bspfile.h#L796)
[^L1]: 2013 Source SDK `dleaf_t` [`src/public.bspfile.h` line 823](https://github.com/ValveSoftware/source-sdk-2013/blob/master/sp/src/public/bspfile.h#L823)

**Comment by sum-catnip** [comment]
My bad, youre completely right. But then im very confused where that "pack expected 18 items for packing (got 17)" is comming from

**Comment by snake-biscuits** [comment]
This prints out the names of each map w/ no errors for me:
```python
>>> import bsp_tool, os, fnmatch
>>> md = "D:/SteamLibrary/steamapps/common/Portal 2/portal2/maps"
>>> maps = {m[:-4]: bsp_tool.load_bsp(os.path.join(md, m)) for m in fnmatch.filter(os.listdir(md), "*.bsp")}
>>> for m, b in maps.items():
...     print(m)
...     try:
...             x = b.lump_as_bytes("LEAVES")
...     except Exception:
...             print("!!! FAILED")
...
```
If you know a particularly error prone map name that'd help with testing on my end, but it seems to be working rn for me

**Comment by snake-biscuits** [comment]
> im very confused where that "pack expected 18 items for packing (got 17)" is comming from

me too
very odd that it's short by 1 in each case (18 & 17, 15 & 14)

**Comment by sum-catnip** [comment]
Oh the error wasnt in leaves. It was in TEXTURE_INFO

**Comment by snake-biscuits** [comment]
ok yeah most maps failed to save TextureInfo to bytes
now that I'm replicating it locally I can dig down and figure this out

**Comment by sum-catnip** [comment]
when im home ill look into which map breaks the leaves. it defs was one of the maps that come with portal2. ill be more specific in the future, sorryy

**Comment by sum-catnip** [comment]
the output above came from `e1912.bsp` but alot of them fail on those 4-5 lumps.
i wrote a little script to test all the maps:

```python
import sys
from pathlib import Path

import bsp_tool

for path in Path(sys.argv[1]).rglob('*.bsp'):
    bsp = bsp_tool.load_bsp(str(path))
    for k, v in bsp.headers.items():
        try: bsp.lump_as_bytes(k)
        except Exception as e:
            print(f'{path.name}/{k}: {e}')
```
> python script.py ~/.local/share/Steam/steamapps/common/Portal\ 2/portal2/maps/

**Comment by snake-biscuits** [comment]
Saving `e1912.bsp` on the latest commit raised no errors, but the `VISIBILITY` lump definitely got mutated
Not 100% sure what changed, I should probably try to load the map in game...

I'll probably need to get back onto updating `extensions.diff` to properly debug if it's causing issues
Looks like minor implementation details at a glance though

PVS & PAS offsets are different (same data different order?)

Some spots in the data might also be using `0XFF` for clusters "visible from everywhere"
`bsp_tool` is precise to `num_clusters`

tl;dr: might be fixed on my machine, need to test more maps though

**Comment by snake-biscuits** [comment]
This script exported every map with no errors:
```python
>>> import bsp_tool, os, fnmatch
>>> src = "D:/SteamLibrary/steamapps/common/Portal 2/portal2/maps"
>>> dst = "E:/Mod/Portal2/save_as_testing"
>>> maps = {m: bsp_tool.load_bsp(os.path.join(src, m)) for m in fnmatch.filter(os.listdir(src), "*.bsp")}
>>> for map_name, bsp in maps.items():
...     print(f"saving {map_name}")
...     bsp.save_as(os.path.join(dst, map_name))
```

Would have to do some kind of Quality Control:
 * update `extensions.diff` so `diff.bsps.BspDiff.what_changed` actually spits out lump names
 * `extensions.diff.id_software.quake2.VisibilityDiff` to see if the changes are equivalent
 * load up maps in-game to check for bugs
 
Also, when I checked `e1912` yesterday, I used HxD which doesn't show all changes, you find them as you scroll
So some other lumps might've changed
Lump headers were a 1:1 match though, which bodes well

**Comment by snake-biscuits** [comment]
> create `extensions.diff.id_software.quake2.VisibilityDiff` to see if the changes are equivalent

done in https://github.com/snake-biscuits/bsp_tool/commit/1c162550e5a9d86022a5e640143f22d36cd5916c

ran against the saved `.bsp`s in the last comment, everything appears to be equivalent

**Comment by snake-biscuits** [comment]
testing saved maps on my machine they seem to work fine
@sum-catnip  please let me know if anything is still breaking on your machine so I can help / close this issue

**Comment by sum-catnip** [comment]
Ill try it tomorrow, thank youu :)

**Comment by sum-catnip** [comment]
hey, i made sure to get the newest bsptool version with `https://github.com/snake-biscuits/bsp_tool.git`. then i tried the script again with a fresh portal2 installation. i even verified the gamefiles in steam. but for some reason almost every map failed in a simmilar way:
```
e1912.bsp/TEXTURE_INFO: pack expected 18 items for packing (got 17)
e1912.bsp/LEAVES: pack expected 15 items for packing (got 14)
e1912.bsp/DISPLACEMENT_INFO: ushort format requires 0 <= number <= 65535
e1912.bsp/GAME_LUMP: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```
i also ran your script and got this:
```
PS Z:\portal2> python new.py
saving e1912.bsp
Failed to convert TEXTURE_INFO to bytes!
Traceback (most recent call last):
  File "Z:\portal2\new.py", line 7, in <module>
    bsp.save_as(os.path.join(dst, map_name))
  File "C:\Users\catni\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\bsp_tool\valve.py", line 137, in save_as
    raise exc
  File "C:\Users\catni\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\bsp_tool\valve.py", line 134, in save_as
    lump_bytes = self.lump_as_bytes(LUMP.name)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\catni\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\bsp_tool\valve.py", line 116, in lump_as_bytes
    raw_lump = b"".join([struct.pack(_format, *x.as_tuple()) for x in lump_entries])
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\catni\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\bsp_tool\valve.py", line 116, in <listcomp>
    raw_lump = b"".join([struct.pack(_format, *x.as_tuple()) for x in lump_entries])
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: pack expected 18 items for packing (got 17)⏎
```

**Comment by snake-biscuits** [comment]
OH! Python 3.11
that is currently broken for some reason
I forgot to check that

afaik something changed with the builtin struct module
try Python 3.10 if you can

**Comment by sum-catnip** [comment]
ooooh, than that must have been the issue haha. ill try with python 3.10 soon but i have a feeling that'll fix the problem :p

**Comment by snake-biscuits** [comment]
I'll get onto debugging 3.11 anyway in #119
Totally forgot I didn't make an issue for that

**Comment by sum-catnip** [comment]
they all save in python3.10 :)
thank you for all the amazing work you do!
